test/fft.cpp

```diff
@@ -682,82 +686,232 @@ TEST(fft3, GFOR) {
+};
+
+class FFTBase : public ::testing::TestWithParam<fft_params> {};
+
+class FFTC2R2D : public FFTBase {};
+class FFT2D : public FFTBase {};
+class FFTC2R3D : public FFTBase {};
+class FFT3D : public FFTBase {};
+class FFTC2R : public FFTBase {};
+class FFTND : public FFTBase {};
+
+string to_test_params(const ::testing::TestParamInfo<FFTBase::ParamType> info) {
+    stringstream ss;
+    ss << "d0_" << info.param.input_dims_[0] << "_d1_"
+       << info.param.input_dims_[1] << "_d2_" << info.param.input_dims_[2]
+       << "_d3_" << info.param.input_dims_[3] << "_"
+       << ((info.param.is_odd_) ? string("odd") : string("even")) << "_norm_"
+       << info.param.norm_factor_;
+    string out = ss.str();
+    return out.replace(out.find("."), 1, "_");
 }
 
-TEST(fft2, InPlace) {
-    array a = randu(1024, 1024, c32);
-    array b = fft2(a);
-    fft2InPlace(a);
+INSTANTIATE_TEST_CASE_P(Inputs2D, FFTC2R2D,
+                        ::testing::Values(
+                            fft_params(dim4(513, 512), false, 0.5),
+                            fft_params(dim4(1025, 1024), false, 0.5),
+                            fft_params(dim4(2049, 2048), false, 0.5)
+                            ),
+                        to_test_params);
+
+INSTANTIATE_TEST_CASE_P(
+    Inputs2D, FFT2D,
+    ::testing::Values(fft_params(dim4(512, 512), false, 0.5),
+                      fft_params(dim4(1024, 1024), false, 0.5),
+                      fft_params(dim4(2048, 2048), false, 0.5)),
+    to_test_params);
+
+INSTANTIATE_TEST_CASE_P(
+    Inputs3D, FFTC2R3D,
+    ::testing::Values(fft_params(dim4(512, 512, 3), false, 0.5),
+                      fft_params(dim4(1024, 1024, 3), false, 0.5),
+                      fft_params(dim4(2048, 2048, 3), false, 0.5)),
+    to_test_params);
+
+INSTANTIATE_TEST_CASE_P(Inputs3D, FFT3D,
+                        ::testing::Values(
+                            fft_params(dim4(1024, 1024, 3), true, 0.5),
+                            fft_params(dim4(1024, 1024, 3), false, 0.5)),
+                        to_test_params);
+
+
+INSTANTIATE_TEST_CASE_P(InputsND, FFTND,
+                        ::testing::Values(
+                            fft_params(dim4(512), false, 0.5),
+                            fft_params(dim4(1024), false, 0.5),
+                            fft_params(dim4(1024, 1024), false, 0.5),
+                            fft_params(dim4(1024, 1024, 3), false, 0.5)),
+                        to_test_params);
+
+
+INSTANTIATE_TEST_CASE_P(InputsND, FFTC2R,
+                        ::testing::Values(
+                            fft_params(dim4(513), false, 0.5),
+                            fft_params(dim4(1025), false, 0.5),
+                            fft_params(dim4(1025, 1024), false, 0.5),
+                            fft_params(dim4(1025, 1024, 3), false, 0.5)),
+                        to_test_params);
+
+// Does not work well with CUDA 10.1
+// TEST_P(FFTC2R2D, Complex32ToRealInputsPreserved) {
```

Comment:
Not a major requirement, we should lean towards using `DISABLED_` prefix for the test so that it is reported as disabled in ctest suite runs.